rust/arrow/src/array/builder.rs

```diff
@@ -500,6 +500,49 @@ impl<T: ArrowPrimitiveType> PrimitiveBuilder<T> {
         Ok(())
     }
 
+    /// Appends values from a slice of type `T` and a validity byte slice
+    pub fn append_values(
+        &mut self,
+        values: &[T::Native],
+        is_valid: &[u8],
+        offset: usize,
+    ) -> Result<()> {
+        let ceil = bit_util::round_upto_power_of_2(values.len(), 8);
+        if offset == 0 {
+            // if there is no offset, the value slice should align with validity
+            assert_eq!(
+                ceil / 8,
+                is_valid.len(),
+                "value slice not aligned with validity slice"
+            );
+        } else {
+            // when there is an offset, the validity slots should be > values
+            assert!(
+                ceil / 8 <= is_valid.len(),
+                "value slots greater than validity slots"
+            );
+        }
+        let mut offset = offset;
+        let mut bools = Vec::with_capacity(values.len() / 8 + 1);
+        is_valid.iter().for_each(|v| {
+            if offset >= 8 {
+                offset -= 8;
+                return;
+            }
+            let bin = format!("{:b}", v);
+            // TODO: reversing to correct the endianness, find a better solution
+            // if we have `11001010` it means first value is null, and 8th value is valid
+            // however if the offset = 3, we should have `11001`, so we skip offset
+            bin.as_str().chars().rev().skip(offset).for_each(|c| {
+                bools.push(c == '1');
+            });
+            // offset should be covered by available bytes
+            offset = 0;
+        });
+        self.bitmap_builder.append_slice(&bools[0..values.len()])?;
```

Comment:
it might be faster to manually call `bimap_builder.reserve(values.len())` ahead of time, then call `bitmap_builder.append()` in the loop to avoid having to create bools vector. `append_slice` is just `append` running in a loop, so if you are already looping manually before calling it, then I feel like using `append` would be better.